### DISCUSSION TITLE
some comparisons, confusions, & praisals

### DISCUSSION THREAD

**Discussion by goyalyashpal** [c3lang/c3c#1102] [Q&A] [answered]
hi there! 

i have known about c3, c2 before but ignored lest can't track innumerably numbered proliferations; 
but recently, the mention in one of [video from zig showtime][vzs] brought it into my solid attention.

[vzs]: https://www.youtube.com/watch?v=1Xdi5jSCYnc&t=1m03s "Loris Cro on Handmade Seattle Bans Language Talks"

[site]: https://c3-lang.org/compare/#zig
[permalink]: https://github.com/c3lang/c3docs/blame/715c30ab03574f1efd65d7147249a5ecd56f0aca/docs/compare.md#L96-L120 "In C3 but not in Zig"

[zig build sys]: https://www.youtube.com/watch?v=Gv2I7qTux7g&t=37m22s
[c abi]: https://www.youtube.com/watch?v=Gv2I7qTux7g&t=2m55s

so, with respect to comparisons ([site] | [permalink]), i had some questions:

* the zig comparisons seem to be from 2-4 yo; and 
* it says these are _"not in Zig"_: `- Integrated build system.   - C ABI compatibility by default.`
* but with reference to this talk from 2019.05 at "ChariotSolutions" titled "The Road to Zig 1.0 - Andrew Kelley"
    - zig has a build system; (ref timestamp: [zig build sys])
    - maintaining C ABI compatibility was one important founding reasons 
    - so, i'd assume that they maintain it (though i don't understand ABI, so, don't know for sure). (ref timestamp: [c abi])
    - edit: oh wait, as per this page: https://c3-lang.org/cinterop/ , i think zig does satisfy this requirement
* any comparisons with c2 language itself? 
as it seems even c2 is evolving: http://www.c2lang.org - like auto-arguments, template functions and structs, etc. 
so like, does c3 keeps update with c2, or is that a divergence? other way to put it is, when was last time c3 got updated with c2 changes?

also:
+ the "[Module][modules] system" - is it similar to `package` in java, or `module` in python? 
edit: oh, i guess the inspiration comes more from cpp custom namespaces) seems interesting, nice :)
+ the "[contracts]" are nice - the optional evaluation based on [compile mode] is also good.
pre-/post-conditions, `@pure` (wow), ada-like `@in` - `@out` (wow)
+ this compile mode stuff and [testing] both feel similar to zig 😃
+ i didn't know "[dynamic interface]" so, thanks for their documentation

[compile mode]: https://c3-lang.org/allfeatures/#safe-fast
[testing]: https://c3-lang.org/allfeatures/#testing
[modules]: https://c3-lang.org/modules/
[contracts]: https://c3-lang.org/contracts/
[dynamic interface]: https://c3-lang.org/anyinterfaces/#dynamic-methods "anyinterfaces/#dynamic-methods"

**Comment by lerno** [comment, marked as answer]
@goyalyashpal The Zig build system is not integrated. Maybe it's unclear what I mean. Zig has one (or more) external build scripts, written in Zig. The advantage of the Zig scheme, is that it offers great flexibility at compile time. The downside is that for an IDE to execute the same things that the build script does, it needs to understand the code.

So the way the C3 compiler works is that it takes some declarations and from those it builds the library or executable. There is no declaration in the project settings that isn't available at the command line, so that you can either choose to have it integrate with a system that also does dependency management, such as CMake, or – for simpler projects – just use "project.json" which is a configuration file. 

This is why I write "Toolchain uses build files written in native Zig" for Zig and "Integrated build system" for C3.

> maintaining C ABI compatibility was one important founding reasons

Zig doesn't have C ABI compatibility for its features. Only a limited subset can be used to interface with C. Normal zig structs, tagged unions, errors etc may not be used in the C interface. In contrast, all structs, unions, enums, optionals etc in C3 are directly available from C.

It is more proper to say that Zig has a FFI to C. Zig also offers a way to import C headers and have those implicitly translated, but that is not related to the ABI.

> any comparisons with c2 language itself?

Auto-arguments are fairly new I think, and is a result of needing some macro definitions not available in the new C2 compiler in development. Template functions and structs are not implemented yet, nor is there a description of how they would look, so it's kind of early to compare with it. The last time when C3 got updated with C2 features was in 2019.

> the [Module](https://c3-lang.org/modules/) system (is it similar to package in java, or module in python?

It is something of a remix. As you already noted, it has a similar behaviour as namespaces, but there are things such as path shortening that doesn't have anything corresponding to it in any other language, but in practice invoking functions from other modules may appear similar to invoking static methods on Java classes, (e.g. you would invoke `fprintf` from `std::io` by writing `io::fprintf(...)` rather than `std::io::fprintf(...)` – although the latter is also allowed)

**Reply by goyalyashpal** [reply]
> _"last time when C3 got updated with C2 features was in 2019."_

Library management (2020-02-14 by Bas vd Berg)

and anything about this? 😃


(i havent searched for this in c3 doc, i am very sorrg for that. but i have got internet for only short duration of time, i couldnt search it myself. i hope u wouldnt mind)

**Reply by lerno** [reply]
The C3 library system was independently developed.

**Reply by goyalyashpal** [reply]
akay. awesome.

**Comment by goyalyashpal** [comment]
hey, also i noticed that the feature of c3 containing <tt>@[test]</tt> and `@benchmark` is not highlighted (like here: https://c3-lang.org/#features or here: https://c3.handmade.network/)

umh, is it so trivial thing?

[test]: https://c3-lang.org/allfeatures/#testing


also, is c3 taking some inspirations from zig? like `defer` or maybe some other things etc :smiley:

edit for contracts:
this at-annotation is awesome way to accomplish this - i have always wanted these contracts and wondered how can this be done - i always stucked on keywords before function definition but that becomes  long-ugly very fast, these `@` annotations in docstring are awesome. (forgive my usage of java parlance)

contracts like the pre/post condition, purity, in/out for ptrs (ada has this too)
i might start with c3 just for these contracts lol. now i havent made any big progs yet, but these things will help me with small things only after which i would be able to quickly move to bigger things and appreciate other features then too

**Reply by lerno** [reply]
Yes, test and benchmark are small things, very useful but not really a huge *language* thing.

> Is c3 taking some inspirations from zig

No, `defer` comes by way of Swift and Jai. I did look extensively at Zig early on and even tried some of its features in C3. They were all removed because I found them deeply problematic in practice: untyped literals combined with implicit widening, no unsigned overflow, special operators for wrapping arithmetics, immediately handled errors. They all led to highly undesirable behaviour. Compile time explicitness in C3 is the opposite of Zig's implicit compile time etc.

Defer is fine, but isn't an invention of Zig.

**Reply by goyalyashpal** [reply]
u had already answered dear 😃 

https://github.com/c3lang/c3c/discussions/1102#discussioncomment-8131761

**Comment by lerno** [comment]
Yes, benchmark and tests all feel like a fairly small feature.

> also, is c3 taking some inspirations from zig? like defer or maybe some other things etc 

`defer` predates Zig by quite a bit, Zig `defer` comes from Jai, which is probably where I got it. That or Swift. A lot of languages have that feature.

As inspiration from Zig, I do indeed look at Zig to see what's added. However, the few things I tried out early on that I "borrowed" from Zig early on (no unsigned wrap, wrapping operators, untyped numerical constants) turned out so badly that I removed them - and in fact went in the opposite direction. It's been useful to see how the direction Zig chose turned out though, that Zig explored the design space has been useful.

Design philosophy wise C3 and Zig are in different places.

**Reply by amarz45** [reply]
> As inspiration from Zig, I do indeed look at Zig to see what's added. However, the few things I tried out early on that I "borrowed" from Zig early on (no unsigned wrap, wrapping operators, untyped numerical constants) turned out so badly that I removed them - and in fact went in the opposite direction.

Why are untyped numerical constants bad? They seem like a good thing on the surface.

**Reply by lerno** [reply]
I mean numerical literals. 

And that's a very long story. But first of all you need to have bidirectional type checking. Then you need to handle things like:
```c
bool foo = (a > 2 ? 1 : 2) == (b > 2 ? 1 : 2);
```
Which poses a problem since the actual type cannot be inferred from anywhere. Compile time folding of bit ops becomes unclear, since for example bit negate is now undefined. (Consider `~0`). And that is not counting the behaviour when interacting with implicit widening and such things.

If the language had no implicit conversions then untyped literals is almost a must, but also the special cases mostly go away. With implicit conversions on top of untyped literals, the unpredictability just increases manyfold with extremely little benefit.

**Comment by LukyGuyLucky** [comment]
With all respect to c3,I like many of its features,may I ask how c3 is very like/dislike the v language(vlang.io)? I get familar to v for a couple of month before knowing c3 thus have an impression they are alike each other,for example,the generics,the operator overloading,the error handling,the Optional/Result ( c3c provided more options which I don't know whether it should be or it is a bit more complicated) ,the way to interop with C etc....

As a programming beginner,if the impression is totally wrong,kindly ignore.

**Reply by lerno** [reply]
Vlang is in some regards trying to write a C-like language but without explicit memory management and a more script-like syntax, borrowing from languages like Go etc.

C3 is essentially C but with syntax tweaks and adding features which are already available as C extensions in some compilers. And then on top of that C3 expands C with semantic macros, contracts, methods, error handling, modules and generic modules.

So perhaps what you recognize as them having things in common, is probably more due that they share C as a common ancestor, although C3 does so much, much more closely.

**Reply by LukyGuyLucky** [reply]
Thank you for the feedback!
